[c10d][gloo] Enable using c10::Half for gloo

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153862

Testing with https://github.com/pytorch/gloo/pull/446 and we see that the numerical issues reported in https://github.com/pytorch/pytorch/issues/152300 is indeed resolved and we added a unit test for it. Also update submodule gloo to reflect the change on the gloo side.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

fduwjj committed:
[c10d][gloo] Enable using c10::Half for gloo

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Enable using c10::Half for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k reviewed (approved):
Change seems reasonable to me, make sure to land the Gloo change first and then update this PR to point to the commit on main instead of the branch

d4l3k commented on test_gloo.py:
```diff
@@ -0,0 +1,30 @@
+import torch
```
probably shouldn't land this here

fduwjj replied:
opps, sorry I should have it removed..

d4l3k commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -465,6 +465,20 @@ def _test_allreduce_basics(self, fn):
         )
 
+        # Single input tests
+        tensor = (torch.rand(20000, 1, dtype=torch.float32) * 2 - 1) * 65504 / self.world_size
```
the np.uniform is a better test than this since it tests every possible bit combination where this only tests the valid fp32 ones (no nan, inf etc)

fduwjj replied:
sure, let me use the numpy one.

fduwjj committed:
Update on "[c10d][gloo] Enable using c10::Half for gloo"


Testing with https://github.com/pytorch/gloo/pull/446 and we see that the numerical issues reported in https://github.com/pytorch/pytorch/issues/152300 is indeed resolved and we added a unit test for it.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Enable using c10::Half for gloo"


Testing with https://github.com/pytorch/gloo/pull/446 and we see that the numerical issues reported in https://github.com/pytorch/pytorch/issues/152300 is indeed resolved and we added a unit test for it.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Enable using c10::Half for gloo"


Testing with https://github.com/pytorch/gloo/pull/446 and we see that the numerical issues reported in https://github.com/pytorch/pytorch/issues/152300 is indeed resolved and we added a unit test for it.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@malfet can you kindly review this PR? Thanks We updated the gloo submodule. Thanks

clee2000 reviewed (approved):
stamping on behalf of tristan, who should have sufficient permissions to approve (add yourself to pytorch/metamates)

malfet reviewed (approved):
LGTM
